File: android/src/main/kotlin/com/navideck/universal_ble/UniversalBleFilterUtil.kt
package com.navideck.universal_ble

import android.annotation.SuppressLint
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.os.ParcelUuid
import android.util.Log
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import kotlin.experimental.and

private const val TAG = "UniversalBlePlugin"

@SuppressLint("MissingPermission")
class UniversalBleFilterUtil {
    var scanFilter: UniversalScanFilter? = null
    var serviceFilterUUIDS: List<UUID> = emptyList()

    fun filterDevice(
        name: String?,
        manufacturerDataList: List<UniversalManufacturerData>,
        serviceUuids: Array<UUID>,
    ): Boolean {
        val filter = scanFilter ?: return true

        val hasNamePrefixFilter = filter.withNamePrefix.isNotEmpty()
        val hasServiceFilter = filter.withServices.isNotEmpty()
        val hasManufacturerDataFilter = filter.withManufacturerData.isNotEmpty()

        // If there is no filter at all, then allow device
        if (!hasNamePrefixFilter &&
            !hasServiceFilter &&
            !hasManufacturerDataFilter
        ) {
            return true
        }

        // For not, we only have DeviceName filter
        return hasNamePrefixFilter && isNameMatchingFilters(filter, name) ||
                hasServiceFilter && isServicesMatchingFilters(serviceUuids) ||
                hasManufacturerDataFilter && isManufacturerDataMatchingFilters(
            filter,
            manufacturerDataList
        )
    }

    private fun isNameMatchingFilters(scanFilter: UniversalScanFilter, name: String?): Boolean {
        val namePrefixFilter = scanFilter.withNamePrefix.filterNotNull()
        if (namePrefixFilter.isEmpty()) {
            return true
        }
        if (name.isNullOrEmpty()) {
            return false
        }
        return namePrefixFilter.any { name.startsWith(it) }
    }

    private fun isServicesMatchingFilters(
        serviceUuids: Array<UUID>,
    ): Boolean {
        if (serviceFilterUUIDS.isEmpty()) {
            return true
        }
        if (serviceUuids.isEmpty()) {
            return false
        }
        return serviceFilterUUIDS.any {
            serviceUuids.contains(it)
        }
    }

    private fun isManufacturerDataMatchingFilters(
        scanFilter: UniversalScanFilter,
        manufacturerDataList: List<UniversalManufacturerData>,
    ): Boolean {
        val filters = scanFilter.withManufacturerData.filterNotNull()
        if (filters.isEmpty()) return true
        if (manufacturerDataList.isEmpty()) return false
        return manufacturerDataList.any { msd ->
            filters.any { filter ->
                msd.companyIdentifier == filter.companyIdentifier &&
                        isDataMatching(filter.data, msd.data, filter.mask)
            }
        }
    }

    private fun isDataMatching(
        filterData: ByteArray?,
        deviceData: ByteArray,
        filterMask: ByteArray?,
    ): Boolean {
        if (filterData == null) return true
        if (filterData.size > deviceData.size) return false

        val mask = filterMask ?: ByteArray(filterData.size) { 0xFF.toByte() }
        if (filterData.size != mask.size) return false

        return filterData.indices.all { i ->
            (filterData[i] and mask[i]) == (deviceData[i] and mask[i])
        }
    }
}

fun UniversalScanFilter.hasCustomFilter(): Boolean {
    // Only NamePrefix Filtering is not allowed in native filters
    return withNamePrefix.isNotEmpty()
}

// Convert UniversalScanFilter to ScanFilter
fun UniversalScanFilter.toScanFilters(serviceUuids: List<UUID>): List<ScanFilter> {
    val scanFilters: ArrayList<ScanFilter> = arrayListOf()

    // Add withServices Filter
    for (service in serviceUuids) {
        try {
            service.let {
                scanFilters.add(
                    ScanFilter.Builder().setServiceUuid(ParcelUuid(it)).build()
                )
                Log.e(TAG, "scanFilters: $it")
            }
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
            throw FlutterError(
                "illegalIllegalArgument",
                "Invalid serviceId: $service",
                e.toString()
            )
        }
    }

    // Add ManufacturerData Filter
    for (manufacturerData in this.withManufacturerData) {
        try {
            manufacturerData?.companyIdentifier?.let {
                val data: ByteArray = manufacturerData.data ?: ByteArray(0)
                val mask: ByteArray? = manufacturerData.mask
                if (mask == null) {
                    scanFilters.add(
                        ScanFilter.Builder().setManufacturerData(
                            it.toInt(), data
                        ).build()
                    )
                } else {
                    scanFilters.add(
                        ScanFilter.Builder().setManufacturerData(
                            it.toInt(), data, mask
                        ).build()
                    )
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
            throw FlutterError(
                "illegalIllegalArgument",
                "Invalid manufacturerData: ${manufacturerData?.companyIdentifier} ${manufacturerData?.data} ${manufacturerData?.mask}",
                e.toString()
            )
        }
    }

    return scanFilters.toList()
}

File: darwin/Classes/UniversalBleFilterUtil.swift
//
//  UniversalBleFilterUtil.swift
//  universal_ble
//
//  Created by Rohit Sangwan on 23/08/24.
//

import CoreBluetooth
import Foundation

public class UniversalBleFilterUtil {
    var scanFilter: UniversalScanFilter?
    var scanFilterServicesUUID: [CBUUID] = []

    func filterDevice(name: String?, manufacturerData: UniversalManufacturerData?, services: [CBUUID]?) -> Bool {
        guard let filter = scanFilter else {
            return true
        }

        let hasNamePrefixFilter = !filter.withNamePrefix.isEmpty
        let hasServiceFilter = !filter.withServices.isEmpty
        let hasManufacturerDataFilter = !filter.withManufacturerData.isEmpty

        // If there is no filter at all, then allow device
        if !hasNamePrefixFilter && !hasServiceFilter && !hasManufacturerDataFilter {
            return true
        }

        // Else check one of the filter passes
        return hasNamePrefixFilter && isNameMatchingFilters(filter: filter, name: name) ||
            hasServiceFilter && isServicesMatchingFilters(services: services) ||
            hasManufacturerDataFilter && isManufacturerDataMatchingFilters(scanFilter: filter, msd: manufacturerData)
    }

    func isNameMatchingFilters(filter: UniversalScanFilter, name: String?) -> Bool {
        let prefixFilters = filter.withNamePrefix.compactMap { $0 }.filter { !$0.isEmpty }

        guard !prefixFilters.isEmpty else {
            return true
        }

        guard let name = name, !name.isEmpty else {
            return false
        }

        return prefixFilters.contains { name.hasPrefix($0) }
    }

    func isServicesMatchingFilters(services: [CBUUID]?) -> Bool {
        let serviceFilters = Set(scanFilterServicesUUID.compactMap { $0 })

        guard !serviceFilters.isEmpty else {
            return true
        }

        guard let services = services, !services.isEmpty else {
            return false
        }

        return !Set(services).isDisjoint(with: serviceFilters)
    }

    func isManufacturerDataMatchingFilters(scanFilter: UniversalScanFilter, msd: UniversalManufacturerData?) -> Bool {
        let filters = scanFilter.withManufacturerData.compactMap { $0 }
        if filters.isEmpty {
            return true
        }

        guard let msd = msd else {
            return false
        }

        for filter in filters {
            let companyIdentifier: Int64 = filter.companyIdentifier
            if msd.companyIdentifier == companyIdentifier && findData(find: filter.data?.toData(), inData: msd.data.toData(), usingMask: filter.mask?.toData()) {
                return true
            }
        }
        return false
    }

    func findData(find: Data?, inData data: Data, usingMask mask: Data?) -> Bool {
        if let find = find {
            // If mask is null, use a default mask of all 1s
            let mask = mask ?? Data(repeating: 0xFF, count: find.count)

            // Ensure find & mask are same length
            guard find.count == mask.count else {
                return false
            }

            for i in 0 ..< find.count {
                // Perform bitwise AND with mask and then compare
                if (find[i] & mask[i]) != (data[i] & mask[i]) {
                    return false
                }
            }
        }
        return true
    }
}

extension UniversalScanFilter {
    var hasCustomFilters: Bool {
        return !withManufacturerData.isEmpty || !withNamePrefix.isEmpty
    }
}
